File: typescript/rekognition-lambda-s3-trigger/lib/rekognition-lambda-s3-trigger-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as iam from 'aws-cdk-lib/aws-iam';
import { AttributeType, Table } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from "constructs";
import * as path from 'path';
import * as s3n from 'aws-cdk-lib/aws-s3-notifications';

export class RekognitionLambdaS3TriggerStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create S3 Bucket
    const bucket = new s3.Bucket(this, 'Bucket', {
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    // create DynamoDB table to hold Rekognition results
    const table = new Table(this, 'Classifications', {
      partitionKey: {
        name: 'image_name',
        type: AttributeType.STRING
      },
      removalPolicy: cdk.RemovalPolicy.DESTROY // removes table on cdk destroy
    });


    // create Lambda function
    const lambdaFunction = new lambda.Function(this, 'RekFunction', {
      handler: 'rekfunction.handler',
      runtime: lambda.Runtime.NODEJS_20_X,
      code: lambda.Code.fromAsset(path.join(__dirname, '../lambda')),
      environment: {
        'BUCKET_NAME': bucket.bucketName,
        'TABLE_NAME': table.tableName
      }
    });

    // add Rekognition permissions for Lambda function
    const statement = new iam.PolicyStatement();
    statement.addActions("rekognition:DetectLabels");
    statement.addResources("*");
    lambdaFunction.addToRolePolicy(statement);

    // create trigger for Lambda function with image type suffixes
    bucket.addEventNotification(s3.EventType.OBJECT_CREATED, new s3n.LambdaDestination(lambdaFunction),{suffix: '.jpg'});
    bucket.addEventNotification(s3.EventType.OBJECT_CREATED, new s3n.LambdaDestination(lambdaFunction),{suffix: '.jpeg'});
    bucket.addEventNotification(s3.EventType.OBJECT_CREATED, new s3n.LambdaDestination(lambdaFunction),{suffix: '.png'});

    // grant permissions for lambda to read/write to DynamoDB table and bucket
    table.grantReadWriteData(lambdaFunction);
    bucket.grantReadWrite(lambdaFunction);
	 new cdk.CfnOutput(this, "UploadImageToS3", {
      value: `aws s3 cp <local-path-to-image> s3://${bucket.bucketName}/`,
      description: "Upload an image to S3 (using AWS CLI) to trigger Rekognition",
    });
    new cdk.CfnOutput(this, "DynamoDBTable", {
      value: table.tableName,
      description: "This is where the image Rekognition results will be stored.",
    });
    new cdk.CfnOutput(this, "LambdaFunction", {
      value: lambdaFunction.functionName,
    });
    new cdk.CfnOutput(this, "LambdaFunctionLogs", {
      value: lambdaFunction.logGroup.logGroupName,
    });
  }
}

File: python/rekognition-lambda-s3-trigger/rekognition_lambda_s3_trigger/rekognition_lambda_s3_trigger_stack.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0
#
# This code is sample only. Not for use in production.
#
# Author: Katreena Mullican
# Contact: mullicak@amazon.com

from aws_cdk import (
    aws_iam as iam,
    aws_lambda as _lambda,
    aws_s3 as s3,
    aws_dynamodb as ddb,
    aws_s3_notifications as s3_notifications,
    Stack
)
from constructs import Construct

class RekognitionLambdaS3TriggerStack(Stack):
    def __init__(self, scope: Construct, id: str, **kwargs) -> None:
        super().__init__(scope, id, **kwargs)

        # create new IAM group and user
        group = iam.Group(self, "RekGroup")
        user = iam.User(self, "RekUser")

        # add IAM user to the new group
        user.add_to_group(group)

        # create S3 bucket to hold images
        # give new user access to the bucket
        bucket = s3.Bucket(self, 'Bucket')
        bucket.grant_read_write(user)

        # create DynamoDB table to hold Rekognition results
        table = ddb.Table(
            self, 'Classifications',
            partition_key=ddb.Attribute(name='image_name', type=ddb.AttributeType.STRING)
        )

        # create Lambda function
        lambda_function = _lambda.Function(
            self, 'RekFunction',
            runtime = _lambda.Runtime.PYTHON_3_8,
            handler = 'rekfunction.handler',
            code = _lambda.Code.from_asset('rekognition_lambda_s3_trigger/lambda'),
            environment = {
                'BUCKET_NAME': bucket.bucket_name,
                'TABLE_NAME': table.table_name
            }
        )

        # add Rekognition permissions for Lambda function
        statement = iam.PolicyStatement()
        statement.add_actions("rekognition:DetectLabels")
        statement.add_resources("*")
        lambda_function.add_to_role_policy(statement)

        # create trigger for Lambda function with image type suffixes
        notification = s3_notifications.LambdaDestination(lambda_function)
        notification.bind(self, bucket)
        bucket.add_object_created_notification(notification, s3.NotificationKeyFilter(suffix='.jpg'))
        bucket.add_object_created_notification(notification, s3.NotificationKeyFilter(suffix='.jpeg'))
        bucket.add_object_created_notification(notification, s3.NotificationKeyFilter(suffix='.png'))

        # grant permissions for lambda to read/write to DynamoDB table and bucket
        table.grant_read_write_data(lambda_function)
        bucket.grant_read_write(lambda_function)
